### ISSUE TITLE
Pub/Sub hardly returns any messages on pull request while there a thousands waiting

### ISSUE THREAD

**Issue by Chasx** [googleapis/google-cloud-dotnet#5000]
--edit--
 - OS is LInux Mint Mate 19.3 64-bit (Ubuntu 18.04 based)
 - dotnet core 2.1 (dotnet-sdk-2.1.802-linux-x64)
 - Google.Cloud.PubSub.V1 (version 2.0.0)

--edit--

 I use Google Pub/Sub as a buffering queue for messages that I regularly poll for processing. One process publishes about 100.000 messages in a certain time frame and a subscriber is started regularly to pull all queued message to process them.

I use C# and the queue is in pull mode, receiving the messages synchronously. The code is about the same as [the example code for SubscriberServiceApiClient](https://googleapis.github.io/google-cloud-dotnet/docs/Google.Cloud.PubSub.V1/)

```
int messagesInQueue = getFromMonitoring();
while (totalProcessed < messagesInQueue) {
  PullResponse response = subscriber.Pull(SUBSCRIPTION_NAME,
                                          returnImmediately: true,
                                          maxMessages: 1000);
  foreach (ReceivedMessage received in response.ReceivedMessages) {
       // process them
  }
  
  if (response.ReceivedMessages.Any())
    subscriber.Acknowledge(subscriptionName, response.ReceivedMessages.Select(m => m.AckId));

  // Console write the number of messages now processed and total processed
}
```

This_ used to _work quite well, I regularly got responses with about 1000 messages which is myspecified 'maxMessages'. Sometimes less but that's ok. I could manage to empty the topic within a reasonable time.

Then I made a change in pub/sub: the publisher still does the same but there are now *2 subscriptions* to the same topic. The new one is a 'push' subscription which automatically handles messages in the topic in a Google Cloud Function. This works fine but the C# counterpart 'pull' subscriber is now *unbearably slow* receiving messages in the topic.

Almost every 'pull' returns with only a handful of messages, even when there are literally thousands waiting:

```
*** Number of items in queue is approximately: [54573] *** 
Processing [29] items (total received: [29])...
Processing [259] items (total received: [288])...
Processing [12] items (total received: [300])...
Processing [18] items (total received: [318])...
Processing [4] items (total received: [322])...

... a lot of moments later ...

Processing  [40] items (total received: [54358])...
Processing  [193] items (total received: [54551])...
Processing  [129] items (total received: [54680])...
Received a total of [54680] messages in [31:35]...
```

As you can see it now takes **more than 30 minutes to empty this queue with ~50.000 messages waiting** which is really bad. Previously this was done in a couple of minutes max.

Does anyone have an explanation/a solution for this?

Note:

- I don't want to use asynchronous mode. Although it solves the problem above because I can run the async process longer and use a cancellation token to stop polling, it does not allow me to ACK the messages outside of the message loop (`subscriber.Acknowledge`) does not exist in async mode). On my side this would cause an enormous performance penalty because I cannot bulk ACK a whole batch of messages anymore; I need to ACK (or NACK) them in the message receive loop one by one so I also need to make sure the processing was completely don. That means a F_SYNC to disk for _every_ message instead of bulk writing, F_SYNC and then ACK-ing all of them (or NACK-ing all of them if any single message failed, no biggie, it's an exception and they will then all return to the queue, PubSub is not 'exactly once delivery' anyway).

- Setting `returnImmediately` to false does not change this behavior. It only makes the process never return 0 messages but at least 1. As there are already 1000's of messages waiting it doesn't matter.

- I tried using the `SubscriberServiceApiClientBuilder` to use a callsettings object with `CallSettings.FromExpiration(Expiration.FromTimeout(MY_TIMEOUT))` because I hoped that using a timeout would make the `pull` not return as fast but it does not change the behavior either. Having a mechanism to just letting the call wait a little longer before returning would be a fine solution instead of returning fast with only so little messages.

**Comment by jskeet** [comment]
Thanks for raising this issue. I'll assign it to Chris to start with as he's done more work on PubSub than I have, but I *suspect* we'll need to ask the PubSub team about this. I very much doubt that this is something the client library itself will be able to fix - especially as you're not even using the higher level API (where more manually-written code is involve).

Just to check that I've really understood the situation in a "TL;DR" fashion I can present to start with: 

- With only a single subscription to the topic, you're able to pull "roughly 1000" messages at a time
- Just adding another subscription to the same topic (and it shouldn't really matter what that subscription is doing), you're only able to pull a few hundred per call.

Have I understood correctly?

**Comment by Chasx** [comment]
Yes that is correct, but after some more testing I am not 100% sure any more if adding the extra subscription is the root of the cause. It could also have something to do with the latency when doing the 'pull' but this is still a hunch.

When there was only 1 subscription I called the 'pull' method over anasymmetric VPN (25/5 mbps) line. Now I 'pull' directly on my 100/100 fiber connection and the responses _seem_ to be quicker, but also contain less messages. I don't think that 'my speed' would effect the number of messages to return but maybe the call contains timestamps so PubSub optimizes the number of returned messages based on turn around time?

Still; I would expect a single call to (mostly) return more than 50% of the requested 'maxMessages' and not just 10 or 20. Especially if this is the 'x'th subsequent pull to the queue.

**Comment by jskeet** [comment]
Right, thanks for the extra information. If you're able to easily perform multiple tests with and without the extra subscription, keeping everything else the same, that would be really helpful. (e.g. 5 with two subscriptions, 5 with just the one subscription).

If I were to try to reproduce this in a naive way, I would:

- Create a new topic
- Create a subscription to it
- Publish 100,000 messages
- Pull all of those messages from the subscription using the code you've shown, recording how many I get each time

As far as you're aware, is that a reasonable reproduction of your system?

**Comment by Chasx** [comment]
Yes it is. Because the topic is 'in production' I cannot just stop the current push subscriber to see if the 'pull' behaviour returns to normal.

But the coming days I could create a simple project to support my findings if it would help?

**Comment by jskeet** [comment]
That would be really helpful if you're able to do that - on Monday I can try to do it as well, and we can compare findings. If you want to wait until after I've tried to reproduce it, that's fine - I do often find that multiple reproduction attempts in parallel can be useful, just to avoid making assumptions about the environment, but I understand that it does mean potentially redundant effort.

**Comment by Chasx** [comment]
No problem, happy to help. I will create some test code this weekend and post back.

**Comment by Chasx** [comment]
It is quite easy to trigger the problem. I does not need a secondary subscriber as described before, but maybe the amounts of messages pulled get worse when you add one (haven't tested it).

The console test program creates a publisher that puts 3.000 items in the queue, each one 100 ms apart, so it takes about 5 minutes to complete. Then the subscriber tries to pull the messages. Ideally it would take only 3 'pulls' because the maximum messages to pull is set to 1000. Instead my test run takes 16 'pulls' to receive all messages:

```
Published: [3000 of 3000] messages    
...
[775] new messages pulled in [3607] ms. and acknowledged in [54] ms. Total pulled [775] out of [3000]) 
[249] new messages pulled in [1201] ms. and acknowledged in [40] ms. Total pulled [1024] out of [3000]) 
[38] new messages pulled in [1982] ms. and acknowledged in [20] ms. Total pulled [1062] out of [3000]) 
[25] new messages pulled in [1130] ms. and acknowledged in [23] ms. Total pulled [1087] out of [3000]) 
[20] new messages pulled in [1829] ms. and acknowledged in [18] ms. Total pulled [1107] out of [3000]) 
[14] new messages pulled in [1136] ms. and acknowledged in [19] ms. Total pulled [1121] out of [3000]) 
[25] new messages pulled in [1833] ms. and acknowledged in [20] ms. Total pulled [1146] out of [3000]) 
[174] new messages pulled in [1141] ms. and acknowledged in [38] ms. Total pulled [1320] out of [3000]) 
[317] new messages pulled in [2268] ms. and acknowledged in [32] ms. Total pulled [1637] out of [3000]) 
[215] new messages pulled in [2123] ms. and acknowledged in [37] ms. Total pulled [1852] out of [3000]) 
[143] new messages pulled in [1168] ms. and acknowledged in [30] ms. Total pulled [1995] out of [3000]) 
[0] new messages pulled in [2112] ms. and acknowledged in [0] ms. Total pulled [1995] out of [3000]) 
[350] new messages pulled in [1970] ms. and acknowledged in [31] ms. Total pulled [2345] out of [3000]) 
[240] new messages pulled in [1169] ms. and acknowledged in [29] ms. Total pulled [2585] out of [3000]) 
[386] new messages pulled in [2090] ms. and acknowledged in [36] ms. Total pulled [2971] out of [3000]) 
[29] new messages pulled in [1134] ms. and acknowledged in [11] ms. Total pulled [3000] out of [3000]) 
````
Obviously this is just one case. Running the program again gives different results. But in this simple test I regularly pull only tens (once even 0) messages although the calls take ~2 seconds to return...

The code for the test program is printed below. It needs a valid GOOGLE_PROJECT_ID to run (and the environment setting to the credentials), all the rest is done in code.

```
using System;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using Google.Cloud.PubSub.V1;
using Google.Protobuf;

namespace PubSubProblem
{
    class Program
    {
        private const string GOOGLE_PROJECT_ID = ;
        private const string GOOGLE_TOPIC_ID = "testtopic";
        private const string GOOGLE_PULL_SUBSCRIPTION_ID = "testtopic_pull_subscription";
        private const int GOOGLE_MAX_MESSAGES_TO_FETCH_PER_PULL = 1000;
        private const int GOOGLE_ACK_DEADLINE = 600; // NOTE: Default is 60!

        private const int TOTAL_MESSAGES_TO_PUBLISH = 3000;
        private const int SLEEP_BETWEEN_PUBLICATIONS = 100; // ms. to sleep between each 'publish' call

        private static void setupPubSubTopic()
        {
            PublisherServiceApiClient publisherService = PublisherServiceApiClient.Create();
            TopicName topicName = new TopicName(GOOGLE_PROJECT_ID, GOOGLE_TOPIC_ID);
            publisherService.CreateTopic(topicName);

            SubscriberServiceApiClient subscriber = SubscriberServiceApiClient.Create();
            SubscriptionName subscriptionName = new SubscriptionName(GOOGLE_PROJECT_ID, GOOGLE_PULL_SUBSCRIPTION_ID);
            subscriber.CreateSubscription(subscriptionName, topicName, pushConfig: null, ackDeadlineSeconds: GOOGLE_ACK_DEADLINE);
        }

        private static void publish()
        {
            PublisherServiceApiClient publisherService = PublisherServiceApiClient.Create();
            TopicName topicName = new TopicName(GOOGLE_PROJECT_ID, GOOGLE_TOPIC_ID);

            for (int i = 1; i <= TOTAL_MESSAGES_TO_PUBLISH; i++)
            {
                string messageText = $"This is message [{i}] of [{TOTAL_MESSAGES_TO_PUBLISH}]";
                PubsubMessage message = new PubsubMessage {Data = ByteString.CopyFromUtf8(messageText)};
                PublishResponse r = publisherService.Publish(topicName, new[] {message});
                Console.Write($"Published: [{i} of {TOTAL_MESSAGES_TO_PUBLISH}] messages   \r");

                Thread.Sleep(SLEEP_BETWEEN_PUBLICATIONS);
            }

            Console.WriteLine("\n...");
        }

        private static void pull()
        {
            int totalPulled = 0;
            SubscriberServiceApiClient subscriber = SubscriberServiceApiClient.Create();
            SubscriptionName subscriptionName = new SubscriptionName(GOOGLE_PROJECT_ID, GOOGLE_PULL_SUBSCRIPTION_ID);

            while (totalPulled < TOTAL_MESSAGES_TO_PUBLISH)
            {
                Stopwatch stopwatch = new Stopwatch();
                stopwatch.Start();

                PullResponse response = subscriber.Pull(subscriptionName, returnImmediately: true, maxMessages: GOOGLE_MAX_MESSAGES_TO_FETCH_PER_PULL);
                totalPulled += response.ReceivedMessages.Count;
                Console.Write($"[{response.ReceivedMessages.Count}] new messages pulled in [{stopwatch.ElapsedMilliseconds}] ms.");

                stopwatch.Restart();
                if (response.ReceivedMessages.Any())
                    subscriber.Acknowledge(subscriptionName, response.ReceivedMessages.Select(m => m.AckId));

                Console.WriteLine($" and acknowledged in [{stopwatch.ElapsedMilliseconds}] ms. Total pulled [{totalPulled}] out of [{TOTAL_MESSAGES_TO_PUBLISH}]) ");
            }
        }

        private static void deletePubSubTopic()
        {
            SubscriberServiceApiClient subscriber = SubscriberServiceApiClient.Create();
            SubscriptionName subscriptionName = new SubscriptionName(GOOGLE_PROJECT_ID, GOOGLE_PULL_SUBSCRIPTION_ID);
            subscriber.DeleteSubscription(subscriptionName);

            PublisherServiceApiClient publisherService = PublisherServiceApiClient.Create();
            TopicName topicName = new TopicName(GOOGLE_PROJECT_ID, GOOGLE_TOPIC_ID);
            publisherService.DeleteTopic(topicName);
        }

        // ============================================================================================================
        // Do the pubsub tests
        // ============================================================================================================
        static void Main(string[] args)
        {
            setupPubSubTopic();
            publish();
            pull();
            deletePubSubTopic();
        }
    }
}
```

**Comment by jskeet** [comment]
Thanks for the repro - I'll look at it in more detail on Monday.

**Comment by jskeet** [comment]
Okay, I've been looking into for a few hours this morning. Firstly, I can reproduce the behavior you're seeing, which is a good start.

I rewrote the code a little and added a few more options - in particular being able to tweak the publishing batch size, as well as whether the regular `Pull` method is used or `StreamingPull`. (I've provided the code at the bottom.)

I've discovered:

- If the publisher either uses large batches or publishes the messages close together (e.g. removing the 100ms delay between messages) then the subscriber receives more messages at a time
- Using StreamingPull delivers messages with small batches (if the publication batch size is small) but *overall* returns them much quicker

Is it possible that the change in performance you've observed is due to your publisher publishing at a different rate, or batching differently? It would be nice to be able to explain why this has become worse for you.

In terms of improving your throughput, assuming you can't change the publisher, I wonder whether it would be worth you batching locally, either using the high-performance subscriber or using streaming pull to fetch, but only process it in large batches locally. It's *possible* we could provide this out of the box in future versions of the client library, but I wouldn't like to promise that at this point.

```csharp
using Google.Cloud.PubSub.V1;
using Google.Protobuf;
using System;
using System.Diagnostics;
using System.Globalization;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace Issue5000
{
    internal sealed class Program
    {
        private const string ProjectId = "...";
        private const int PublishBatchSize = 1;
        private const int PublishBatchCount = 3000;
        private const int TotalMessages = PublishBatchSize * PublishBatchCount;
        private const int ReportingInterval = 500;
        private static readonly TimeSpan PublishDelay = TimeSpan.FromMilliseconds(1);
        private const int SubscriptionBatchSize = 1000;
        private const bool ReturnImmediately = false;
        // Note: not const, to avoid unreachable code.
        private static readonly bool UseStreamingPull = true;

        private static async Task Main()
        {
            var publisher = PublisherServiceApiClient.Create();
            var subscriber = SubscriberServiceApiClient.Create();
            var runId = DateTime.UtcNow.ToString("yyyyMMddTHHmmss", CultureInfo.InvariantCulture);
            var topicName = new TopicName(ProjectId, $"issue5000-{runId}");
            var subscriptionName = new SubscriptionName(ProjectId, $"issue5000-{runId}");

            publisher.CreateTopic(topicName);
            subscriber.CreateSubscription(subscriptionName, topicName, pushConfig: null, ackDeadlineSeconds: 60);

            PublishMessages(publisher, topicName);
            if (UseStreamingPull)
            {
                await StreamingPullMessages(subscriber, subscriptionName);
            }
            else
            {
                PullMessages(subscriber, subscriptionName);
            }
        }

        private static void PublishMessages(PublisherServiceApiClient publisher, TopicName topicName)
        {
            int published = 0;
            int nextReport = ReportingInterval;
            for (int batch = 1; batch <= PublishBatchCount; batch++)
            {
                var messages = Enumerable.Range(1, PublishBatchSize)
                    .Select(index => $"Message {batch}-{index}")
                    .Select(text => new PubsubMessage { Data = ByteString.CopyFromUtf8(text) });
                var request = new PublishRequest
                {
                    TopicAsTopicName = topicName,
                    Messages = { messages }
                };
                publisher.Publish(request);
                Thread.Sleep(PublishDelay);
                published += PublishBatchSize;
                if (published >= nextReport)
                {
                    Console.WriteLine($"Published {published} out of {TotalMessages} messages");
                    nextReport = published + ReportingInterval;
                }
            }
            Console.WriteLine($"Publishing completed");
        }

        private static void PullMessages(SubscriberServiceApiClient subscriber, SubscriptionName subscriptionName)
        {
            var totalReceived = 0;
            var overallStopwatch = Stopwatch.StartNew();
            while (totalReceived < TotalMessages)
            {
                var stopwatch = Stopwatch.StartNew();
                var messages = subscriber.Pull(subscriptionName, ReturnImmediately, SubscriptionBatchSize).ReceivedMessages;
                var pullTime = stopwatch.ElapsedMilliseconds;
                stopwatch.Reset();
                subscriber.Acknowledge(subscriptionName, messages.Select(message => message.AckId));
                var ackTime = stopwatch.ElapsedMilliseconds;
                totalReceived += messages.Count;
                Console.WriteLine($"Received {messages.Count} messages. Pull time: {pullTime}ms. Ack time: {ackTime}ms. Total received: {totalReceived}.");
            }
            overallStopwatch.Stop();
            Console.WriteLine($"Total time to received {TotalMessages} messages: {overallStopwatch.ElapsedMilliseconds}ms.");
        }

        private static async Task StreamingPullMessages(SubscriberServiceApiClient subscriber, SubscriptionName subscriptionName)
        {
            var stream = subscriber.StreamingPull();
            await stream.WriteAsync(new StreamingPullRequest
            {
                SubscriptionAsSubscriptionName = subscriptionName,
                ClientId = Guid.NewGuid().ToString(),
                StreamAckDeadlineSeconds = 60
            });
            var responses = stream.GetResponseStream();
            var totalReceived = 0;
            var nextReport = ReportingInterval;
            var overallStopwatch = Stopwatch.StartNew();
            while (totalReceived < TotalMessages)
            {
                var stopwatch = Stopwatch.StartNew();
                await responses.MoveNextAsync();
                var pullTime = stopwatch.ElapsedMilliseconds;
                var messages = responses.Current.ReceivedMessages;
                stopwatch.Reset();
                await stream.WriteAsync(new StreamingPullRequest { AckIds = { messages.Select(message => message.AckId) } });
                var ackTime = stopwatch.ElapsedMilliseconds;
                totalReceived += messages.Count;
                if (totalReceived >= nextReport)
                {
                    Console.WriteLine($"Received {totalReceived} out of {TotalMessages} messages");
                    nextReport = totalReceived + ReportingInterval;
                }
            }
            overallStopwatch.Stop();
            Console.WriteLine($"Total time to received {TotalMessages} messages: {overallStopwatch.ElapsedMilliseconds}ms.");
        }
    }
}
```

**Comment by Chasx** [comment]
Thanks for your reply. I will run your code today or tomorrow. 

> Is it possible that the change in performance you've observed is due to your publisher publishing at a different rate, or batching differently? It would be nice to be able to explain why this has become worse for you.

No, the publisher is dumb IOT device that runs a BASH command in a loop (and it has been for a couple of weeks). The loop gathers some information and publishes each result (about 0.6 / second).

![pubsub](https://user-images.githubusercontent.com/9550599/83525352-26953480-a4e5-11ea-9da0-178ca132765c.png)

The behaviour seems to be very erratic. Yesterday I had 40.000 messages that I could fetch with a run having many 'pulls' with > 500 messages at the time and I was done in a couple of minutes. Today I pulled the topic also with about 40.000 messages and many 'pulls' returned with only 1-10 messages and the whole process took more than 30 minutes!

What do you mean with a 'high-performance subscriber'? 

My optimal solution would be to batch as many waiting messages upon receiving and processing them. It would be nice if there was some sort of flag the I can set in the 'pull' to return after 'at least x seconds (or when maxMessages was reached) so the pull process just takes more time to collect (more) results. Or to be able to ACK the messages outside of the async subscriber so I can batch-ACK them in the way I can with the synchronous subscriber.

**Comment by jskeet** [comment]
> What do you mean with a 'high-performance subscriber'?

I meant the SubscriberClient API - but looking back at your reasons for not wanting to use that, I agree it doesn't help directly.

Using the streaming API (as per my `StreamingPullMessages` method) may work for you though. I can imagine you fetching using that, but only acking when you have a complete batch or after a timeout - and then bulk-acking (still via the stream, using the `StreamingPullRequest.AckIds` property).

I believe I understand what you'd like to see in the API, but I just don't think it's available I'm afraid. We could potentially make a batch version of the `SubscriberClient` API which did something similar to what I've just described, but I don't think we'll be able to do that any time soon, partly as the PubSub team wants to keep the library API similar across multiple languages. I'll pass this issue onto them so they're aware of your use case though.

**Comment by Chasx** [comment]
Thank you! I think the `StreamingPull` is a viable alternative in the way you described batching up to a number of messages or up to a timeout and then process and bulk ACK them. 

For now I created a third subscription to my topic so I can start testing this with the original flow of messages from the publisher. It seems to pick up messages a lot faster however `var messages = responses .Current.ReceivedMessages;` has thus far never returned more than 1 message at the time..?

Being a stream it might be obvious that it only returns a single message and never more than one wouldn't it?

**Comment by jskeet** [comment]
> Being a stream it might be obvious that it only returns a single message and never more than one wouldn't it?

It *does* return more than one if the messages were published in a batch - at least in my experience (you can experiment with that using the code I provided). But if your publisher only ever publishes a single message at a time, that won't help much.

**Comment by Chasx** [comment]
I see, the publisher submits only one message at the time, but I assumed that if there are thousands of messages waiting I might end op receiving more then a single one in each `responses.Current.ReceivedMessages . It doesn't but that does not really matter. Your StreamingPull example is already able to pull ~1000 of 'single' messages per second so emptying a queue of hundredths of thousands is done in a matter of minutes. Way better than before.

Thanks for your help. I think the alternative approach is a fine solution so I think we can close the bug report or leave it in a way as 'feature reqest'.

**Comment by jskeet** [comment]
I'm really glad it's helped out. I'm going to close this issue now as I've communicated the feature request to the PubSub team internally, where they're in a much better position to keep track of it.

Thanks very much for the constructive engagement, testing, provision of code, etc. It makes such a difference!

**Comment by Chasx** [comment]
Maybe one more thing that might help out others facing the same issue:

To ACK the messages the example program by Jon uses

```c#
await stream.WriteAsync(new StreamingPullRequest { AckIds = { messages.Select(message => message.AckId) } });
```

To batch process the messages first, I tried putting the AckIds in a list and only on timeout or 'max batch size' call

```c#
await stream.WriteAsync(new StreamingPullRequest { AckIds = { ackIdList });
``` 

This _does not work_ (or at least not for the message _not_ retrieved in the  same `await responses.MoveNextAsync();` loop). I ended up receiving the Ack-ed messages again after a while.

Instead this works:
```c#
var ackIdList = new List<string>();
while ( --- )
    await responses.MoveNextAsync();
    var messages = responses.Current.ReceivedMessages;
    --- process the messages ---
    ackIdList.AddRange(messages.Select(message => message.AckId));   // store the AckIds

    if (--- time to bulk Ack ---) {
        var ackReq = new StreamingPullRequest();
        ackReq.AckIds.AddRange(ackIdList);
        await stream.WriteAsync(ackReq);
    }
}
```

**Comment by jskeet** [comment]
Hmm... thanks very much for the feedback, although I'm surprised by it. That really *should* be okay, given that it's just acking the messages received (and the proto will have been dispatched before messages changes to the next set of messages). I'm glad your workaround worked though.

**Comment by Chasx** [comment]
It really does show strange behaviour. It might or might not Ack some of the messages. When in the `await responses.MoveNextAsync();` loop it usually does but not all of them. First I thought it had to do with the Ack timeout but that does not seem to be the case.

If you want I can open a new bug report and create some code that does and code does not work, but as I am still experimenting with the API I might also just be using it wrong.

**Comment by jskeet** [comment]
I'll create a new issue and look into it when I get a chance. Thanks for raising it.
